### PR TITLE
[Filter] input combi option

### PR DESCRIPTION
**Pull request by jaeyun-jung**
While getting tensor info in caps negotiation, check input combi option and compare info with combined.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

**Review by gichan-jang — APPROVED** [review]
LGTM

**Comment by jaeyun-jung** [comment]
2021-04-30 PR updated
- check index value in combi option and fix possible mem leak

plz see the changes again.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c [hunk 1]
```diff
@@ -1786,23 +1792,24 @@ static gint
...
+  /* release old list */
+  g_list_free (*prop_list1);
+  g_list_free (*prop_list2);
+  *prop_list1 = *prop_list2 = NULL;
 
   for (i = 0; i < num; i++) {
     if (strv[i][0] == 'i') {
-      val = g_ascii_strtoll (&strv[i][1], NULL, 10);
-      *prop_list1 = g_list_append (*prop_list1, GINT_TO_POINTER (val));
+      val = g_ascii_strtoull (&strv[i][1], NULL, 10);
```
Comment (gichan-jang) [inline gst/nnstreamer/tensor_filter/tensor_filter_common.c:1807]: Add validity check such as line 1774
``
if (errno == ERANGE || val >= NNS_TENSOR_SIZE_LIMIT) {
      ml_loge ("Invalid value %s, cannot set combination option.", strv[i])
``

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_filter/tensor_filter_common.c:1807]: added at line 1822

File: gst/nnstreamer/tensor_filter/tensor_filter.c [hunk 2]
```diff
@@ -754,86 +757,91 @@ gst_tensor_filter_configure_tensor (GstTensorFilter * self,
...
-          gst_tensor_filter_compare_tensors (&in_config.info,
-              &prop->input_meta);
-          goto done;
-        }
-      }
-    } else {
-      gst_tensors_info_copy (&prop->input_meta, &in_config.info);
-    }
+  if (!gst_tensors_config_validate (&in_config)) {
+    GST_ERROR_OBJECT (self, "Invalid caps, failed to configure input info.");
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_filter/tensor_filter.c:761]: I wonder whether `gst_tensors_config_validate` really configures input info.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c [hunk 3]
```diff
@@ -2113,11 +2119,50 @@ gst_tensor_filter_common_get_property (GstTensorFilterPrivate * priv,
...
+  if (priv->combi.in_combi_defined) {
+    for (list = priv->combi.in_combi; list != NULL; list = list->next) {
+      i = GPOINTER_TO_UINT (list->data);
+
+      if (i >= in->num_tensors) {
+        nns_loge ("Invalid input index %u, failed to combine info.", i);
+        goto error;
+      }
+
+      gst_tensor_info_copy (&combined->info[idx++], &in->info[i]);
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_filter/tensor_filter_common.c:2146]: How about adding error handling to ensure `idx` is in the valid range.

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 2 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:761]: cannot understand, if dimension and data type is invalid here, gst_tensors_config_validate() will return false.

File: gst/nnstreamer/tensor_filter/tensor_filter_common.c (hunk 3 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_filter/tensor_filter_common.c:2146]: add condition to check max size (line 2148)

File: gst/nnstreamer/tensor_filter/tensor_filter.c (hunk 2 above)
Comment (dongju-chae) [reply gst/nnstreamer/tensor_filter/tensor_filter.c:761]: Ah, my bad. I thought that the log message is not directly related to `gst_tensors_config_validate`. But, it seems to include the overall function behavior.

**Review by dongju-chae — APPROVED** [review]
LGTM

**Pull request merged**